typescript_typeorm_graphql_project/src/graphql/resolver.ts

```diff
@@ -0,0 +1,161 @@
+import { hash, compare } from "bcryptjs";
+import { isEmail, isEmpty, isLength } from "validator";
+import { sign } from "jsonwebtoken";
+import { Post } from "../entities/Post";
+import { User } from "../entities/User";
+import { Request } from "express";
+
+async function createUser({ userInput }, req: Request) {
+  interface ErrorItem {
+    message: string;
+  }
+
+  const errors: ErrorItem[] = [];
+
+  if (!isEmail(userInput.email)) {
+    errors.push({ message: "Email is invalid." });
+  }
+
+  if (
+    isEmpty(userInput.password) ||
+    !isLength(userInput.password, { min: 5 })
+  ) {
+    errors.push({ message: "Password too short!" });
+  }
+
+  if (errors.length > 0) {
+    const error: Error & { data?: any; code?: number } = new Error(
+      "Invalid input."
+    );
+    error.data = errors;
+    error.code = 422;
+    throw error;
+  }
+
+  const existingUser = await User.findOne({
+    where: {
+      email: userInput.email,
+    },
+  });
+
+  if (existingUser) {
+    const error = new Error("User exists already!");
+    throw error;
+  }
+
+  const hashedPw = await hash(userInput.password, 12);
+
+  const user = new User();
+  user.email = userInput.email;
+  user.name = userInput.name;
+  user.password = hashedPw;
+  user.role = userInput.role;
+
+  const createNewUser = await user.save();
+  return { ...createNewUser, _id: createNewUser.id.toString };
+}
+
+async function login({ email, password }) {
+  const user = await User.findOne({
+    where: { email: email },
+  });
+
+  console.log("User details : ", user);
+
+  if (!user) {
+    const error = new Error("User Not Found.") as any;
+    error.code = 401;
```

Comment:
statusCode 401 is required for unauthorized
for not existis you can use 404

Reply:
Updated the error message from 'User Not Found' to 'Not Authenticated